embassy-nrf/src/rng.rs

```diff
@@ -0,0 +1,154 @@
+use core::convert::Infallible;
+use core::marker::PhantomData;
+
+use embassy::interrupt::InterruptExt;
+use embassy::traits;
+use embassy::util::OnDrop;
+use embassy::util::Signal;
+use embassy::util::Unborrow;
+use embassy_extras::unborrow;
+use futures::Future;
+use rand_core::RngCore;
+
+use crate::interrupt;
+use crate::pac;
+use crate::peripherals::RNG;
+
+impl RNG {
+    fn regs() -> &'static pac::rng::RegisterBlock {
+        unsafe { &*pac::RNG::ptr() }
+    }
+}
+
+static NEXT_BYTE: Signal<u8> = Signal::new();
+
+/// A wrapper around an nRF RNG peripheral.
+///
+/// It has a non-blocking API, through `embassy::traits::Rng`, and a blocking api through `rand`.
+pub struct Rng<'d> {
+    irq: interrupt::RNG,
+    phantom: PhantomData<&'d mut RNG>,
+}
+
+impl<'d> Rng<'d> {
+    pub fn new(
+        _rng: impl Unborrow<Target = RNG> + 'd,
+        irq: impl Unborrow<Target = interrupt::RNG> + 'd,
+    ) -> Self {
+        unborrow!(irq);
+
+        let this = Self {
+            irq,
+            phantom: PhantomData,
+        };
+
+        this.stop();
+        this.disable_irq();
+
+        this.irq.set_handler(Self::on_interrupt);
+        this.irq.unpend();
+        this.irq.enable();
+
+        this
+    }
+
+    fn on_interrupt(_: *mut ()) {
+        NEXT_BYTE.signal(RNG::regs().value.read().value().bits());
```

Comment:
If the `*mut [u8]` is a pointer to the buffer passed to `read`, wouldn't it be UB if `mem::forget` was called on the future and the buffer was dropped?

Reply:
Yes. Same applies to futures using DMA: DMA can be started, the future can be `mem::forget`ed and DMA keeps on writing to a possibly-unallocated buf.

The approach we're taking is to simply accept this and tell the user to not leak any future from embassy. Alternatives are using alloc or using something like `embedded-dma` which is really not ergonomic. There's [talks](https://rust-lang.zulipchat.com/#narrow/stream/187312-wg-async-foundations/topic/mem.3A.3Aforget.2C.20borrowed.20buffers.20and.20DMA) of fixing this at the language level by adding some way for a type to declare it's not "leak-safe".

This is the reason most `new` methods are unsafe. Documenting this in depth is TODO.

Reply:
Do we still DropBomb because panic isn't UB?

Reply:
DropBomb/OnDrop trigger on drop, not on leak (`mem::forget`). We can't have UB on future drop. We've deemed acceptable to have UB on future *leak*, not *drop*.

The way to do this is to use OnDrop to safely cleanup. For example `Uarte` requests DMA to stop and spins a bit until it's actually stopped.

Alternatively when that's not possible, DropBomb is acceptable. For example `Qspi` can't stop an in-progress transfer. so DropBomb is used to panic to prevent UB. Alternatively an OnDrop could spin until the *entire transfer* is done. Both options suck, oh well.